Samsung TV can't be turn off after idle period

ejel opened the pull request:
## Description:
When Samsung TV is idle for a period of time after issued a command,
subsequent 'turn_off' command won't turn off the TV. The issue is seen
in Samsung models with websocket as discussed in #12302.

Detailed explanation of the root cause and fix is in the commit message.

**Related issue (if applicable):** fixes #12302 

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

balloob commented on homeassistant/components/media_player/samsungtv.py:
```diff
@@ -155,18 +154,17 @@ def send_key(self, key):
             return
         try:
-            self.get_remote().control(key)
+            with self.get_remote() as remote:
```
Creating and closing a new connection for every key seems expensive. Why not retry the command with a new connection if we get a `ConnectionClosed` ?

ejel replied:
Would it be possible for me to IM with you at your preferred time so I can clarify and avoid lengthy back and forth?

I can see you say that for (web)socket connections to it's a pattern to keep connection open. My question has been why that's the case for components using (web)socket but not for those that use HTTP, despite HTTP also create TCP socket connection underneath.

balloob replied:
Join us in the #dev channel on Discord. 

HTTP is a stateless protocol, WebSocket is not and thus has more overhead.

ejel replied:
I did some more reading on websocket and I think I can see now where the overhead of websocket could be for establishing connection. Establishing websocket connection first require sending/receiving HTTP upgrade request/response before switching to WebSocket connection. Thanks for pointing me out. Learn something new every day.

I have a change that keep the same approach on keeping the connection and retry when ConnectionClosed is raised. This fixes the `turn_off` issue. However, it appears that for `volume_up` and `volume_down` command, there is a separate issue. After some idle time sending these two commands will first result in no response at all. No exception is raised. The subsequent command, however, will raise `BrokenPipeError`. The issue is not a regression from this fix though. To me this appears to be an issue with the underlying library, either `samsungctl` or `websocket-client`. There is a [similar issue](https://github.com/Ape/samsungctl/issues/61) reported in samsungctl.

I can update this PR with the fix for the `turn_off` issue alone and create a separate issue to track the BrokenPipeError issue and tackle that separately (as the latter might require changes in the library). What do you think?

balloob replied:
It should be fairly simple to have a method `sendCommand`:
 - if closed or never opened, open connection
 - try sending command
 - if we get broken pipe error or other exception, set connection to None and call `sendCommand` again ?

balloob replied:
Shouldn't that solve the cases?

ejel replied:
There is also a case where there is no exception raised from connection but TV won't respond. In this case we won't know whether the command succeeded or not. 

I found the issue within samsungctl and working on a way to fix it.

balloob replied:
Can we then close this PR?

ejel replied:
Sorry, I forgot that the fixes for the two issues are independent to each other. I'm updating this PR with the fix for turn_off issue. The fix in `samsungctl` for the no response issue can come in later.

ejel committed:
Samsung TV can't turn off after idle period

When Samsung TV is idle for a period of time after issued a command,
subsequent 'turn_off' command won't turn off the TV. The issue is seen
in Samsung models with websocket as discussed in #12302.

== Reproducible Steps
1. Turn on TV (either via HA or directly).
2. Issue some commands e.g. volume ups / downs.
3. Wait for ~1 minute.
4. Issue turn_off command via HA. TV won't turn off.
5. Issue subsequent turn off commands won't turn off TV still.
6. However, issue some other commands e.g. volume ups / downs multiple
times in a row and then turn_off will turn off the TV.

== Root Cause
The underlying websocket connection opened by samsungctl get closed
after some idle time. There was no retry mechanism so issued commands
would intermittently fail but the subsequent one would succeed when
`_remote` get recreated. With `turn_off()`, however, there is an
additional call to `self.get_remote().close()` which indirectly caused
new connection to be created and then closed immediately. This causes the
component to stuck in failure mode when turn_off command is repeatly
issued.

== The Fix
Recreate the connection and retry the command if connection is closed
to avoid silent failures due to connection closed. Also set `_remote`
to None after calling close() to put it in correct state.

This fix eliminates intermittent command failure and failure mode in
turn_off().
